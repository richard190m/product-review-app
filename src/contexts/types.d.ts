import { ReactNode } from 'react';

// Reviews
export interface ReviewProviderProps {
  children: ReactNode;
}

export interface Review {
  name: string;
  email: string;
  rating: number;
  title: string;
  comment: string;
  date: Date;
}

export interface HandleSubmit {
  // eslint-disable-next-line no-unused-vars
  (review: Review): void;
}

export interface ReviewsContextInterface {
  reviews: Review[];
  submit: HandleSubmit;
}

// Breakpoint
export interface BreakpointProviderProps {
  children: ReactNode;
}

export interface BreakpointContextInterface {
  breakpoint: string;
}
