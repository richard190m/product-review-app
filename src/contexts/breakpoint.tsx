import { BreakpointContextInterface, BreakpointProviderProps } from './types';
import { createContext, useEffect, useState } from 'react';

export const BreakpointContext =
  createContext<BreakpointContextInterface | null>(null);

export const BreakpointProvider = ({ children }: BreakpointProviderProps) => {
  const [breakpoint, setBreakpoint] = useState<string>('');

  const handleBreakpoint = () => {
    if (window.innerWidth < 768) breakpoint != 'xs' && setBreakpoint('xs');
    else if (window.innerWidth < 1024)
      breakpoint != 'sm' && setBreakpoint('sm');
    else if (window.innerWidth < 1440)
      breakpoint != 'md' && setBreakpoint('md');
    else breakpoint != 'lg' && setBreakpoint('lg');
  };

  useEffect(() => {
    handleBreakpoint();

    function handleResize() {
      handleBreakpoint();
    }

    window.addEventListener('resize', handleResize);

    return () => window.removeEventListener('resize', handleResize);
  }, []);

  return (
    <BreakpointContext.Provider value={{ breakpoint }}>
      {children}
    </BreakpointContext.Provider>
  );
};
