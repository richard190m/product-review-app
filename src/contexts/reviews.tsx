import { createContext, useEffect, useState } from 'react';
import {
  HandleSubmit,
  Review,
  ReviewProviderProps,
  ReviewsContextInterface,
} from './types';
import MockReviews from '../mocks/MockReviews';

export const ReviewsContext = createContext<ReviewsContextInterface | null>(
  null,
);

export const ReviewsProvider = ({ children }: ReviewProviderProps) => {
  const [reviews, setReviews] = useState<Review[]>([]);

  const handleSubmit: HandleSubmit = review => {
    setReviews([...reviews, review]);
  };

  useEffect(() => {
    // Load Reviews (mocks)
    setReviews(MockReviews);
  }, []);

  return (
    <ReviewsContext.Provider value={{ reviews, submit: handleSubmit }}>
      {children}
    </ReviewsContext.Provider>
  );
};
