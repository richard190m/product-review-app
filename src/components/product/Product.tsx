import {
  Container,
  DescriptionContainer,
  InRating,
  PlusMinusIcon,
  ProductCard,
  ProductDesciption,
  ProductDesciptionReview,
  Rating,
  ReviewTitleContainer,
  StarMain,
} from './styles';
import {
  L_BREAKPOINT,
  M_BREAKPOINT,
  S_BREAKPOINT,
} from '../../theme/variables';
import React, { useContext, useEffect, useState } from 'react';
import { BreakpointContext } from '../../contexts/breakpoint';
import ImageCard from './image-card';
import MockProduct from '../../mocks/MockProductInfo';
import ReviewForm from '../review-form';
import { ReviewsContext } from '../../contexts/reviews';

const MENUS = ['descripion', 'review'];

const Product = () => {
  const breakpointContext = useContext(BreakpointContext);
  const reviewsContext = useContext(ReviewsContext);
  const [showDescription, setShowDescription] = useState(true);
  const [showReview, setShowReview] = useState(true);
  const { brand, name, description, price } = MockProduct;
  const breakpoint = breakpointContext?.breakpoint || L_BREAKPOINT;
  const productSumRatings =
    reviewsContext?.reviews.reduce(
      (partialSum, review) => partialSum + review.rating,
      0,
    ) || 0;
  const productRating =
    Math.round(
      (productSumRatings / (reviewsContext?.reviews.length || 1)) * 10,
    ) / 10;

  useEffect(() => {
    if (breakpoint === S_BREAKPOINT || breakpoint === M_BREAKPOINT) {
      setShowDescription(false);
      setShowReview(true);
    } else {
      setShowDescription(true);
      setShowReview(true);
    }
  }, [breakpoint]);

  const handleShowMenu = (menu: string) => {
    switch (menu) {
      case MENUS[0]: {
        if (showDescription) {
          setShowDescription(false);
        } else {
          setShowDescription(true);
          setShowReview(false);
        }
        break;
      }
      case MENUS[1]: {
        if (showReview) {
          setShowReview(false);
        } else {
          setShowDescription(false);
          setShowReview(true);
        }
        break;
      }
      default: {
        break;
      }
    }
  };

  return (
    <Container>
      <ImageCard brand={brand.toUpperCase()} alt={name} />
      <ProductCard>
        <ProductDesciptionReview>
          <h1>{name.toUpperCase()}</h1>
          <Rating>
            <StarMain />
            {productRating}
            <InRating>{' /5 starts'}</InRating>
          </Rating>
          {(breakpoint === S_BREAKPOINT || breakpoint === M_BREAKPOINT) && (
            <DescriptionContainer>
              <h3>DESCRIPTION</h3>
              <PlusMinusIcon
                data-testid="plus-minus-description"
                onClick={() => handleShowMenu(MENUS[0])}
              >
                {showDescription ? '-' : '+'}
              </PlusMinusIcon>
            </DescriptionContainer>
          )}
          {showDescription && (
            <div data-testid="description">
              <ProductDesciption>{description}</ProductDesciption>
              <p>{price}</p>
            </div>
          )}
          <ReviewTitleContainer>
            <h3>MAKE A REVIEW</h3>
            {(breakpoint === S_BREAKPOINT || breakpoint === M_BREAKPOINT) && (
              <PlusMinusIcon
                data-testid="plus-minus-review"
                onClick={() => handleShowMenu(MENUS[1])}
              >
                {showReview ? '-' : '+'}
              </PlusMinusIcon>
            )}
          </ReviewTitleContainer>
          <ReviewForm visible={showReview} />
        </ProductDesciptionReview>
      </ProductCard>
    </Container>
  );
};

export default Product;
