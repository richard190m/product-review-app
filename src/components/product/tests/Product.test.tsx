import { fireEvent, render } from '@testing-library/react';
import { BreakpointProvider } from '../../../contexts/breakpoint';
import Product from '../Product';
import { ReviewsProvider } from '../../../contexts/reviews';

const resizeWindow = (width: number, height: number) => {
  global.window.innerWidth = width;
  global.window.innerHeight = height;
  global.window.dispatchEvent(new Event('resize'));
};

describe('Product', () => {
  test('should render correctly', () => {
    const { container } = render(
      <BreakpointProvider>
        <ReviewsProvider>
          <Product />
        </ReviewsProvider>
      </BreakpointProvider>,
    );

    expect(container).toMatchSnapshot();
  });

  test('should show plus and minus icon when breakpoint e small or medium', async () => {
    // Set window width and height to render as mobile
    resizeWindow(800, 300);

    const { getByTestId } = render(
      <BreakpointProvider>
        <ReviewsProvider>
          <Product />
        </ReviewsProvider>
      </BreakpointProvider>,
    );

    expect(getByTestId('plus-minus-description')).toBeInTheDocument();
    expect(getByTestId('plus-minus-review')).toBeInTheDocument();
  });

  test('should show/hide menus when click plus-minus button', async () => {
    // Set window width and height to render as mobile
    resizeWindow(800, 300);

    const { getByTestId, queryByTestId } = render(
      <BreakpointProvider>
        <ReviewsProvider>
          <Product />
        </ReviewsProvider>
      </BreakpointProvider>,
    );

    const buttonDescription = getByTestId('plus-minus-description');
    const buttonReview = getByTestId('plus-minus-review');
    const description = queryByTestId('description');
    const ReviewForm = queryByTestId('review-form');

    // Show Description
    fireEvent.click(buttonDescription);
    expect(description).toBeDefined();
    // Hide Description
    fireEvent.click(buttonDescription);
    expect(description).toBeNull;

    // Show Review Form
    fireEvent.click(buttonReview);
    expect(ReviewForm).toBeDefined();
    // Hide Review Form
    fireEvent.click(buttonReview);
    expect(ReviewForm).not.toBeNull;
  });
});
