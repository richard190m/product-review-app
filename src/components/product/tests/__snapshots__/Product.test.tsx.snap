// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product should render correctly 1`] = `
<div>
  <section
    class="sc-bczRLJ eUMUfx"
  >
    <div
      class="sc-ftvSup jgoIed"
    >
      <img
        alt="New Collection Shoes"
        class="sc-papXJ kGzhVz"
        src="https://source.unsplash.com/random/700x700/?shoes"
      />
      <span
        class="sc-jqUVSM eiAGhe"
      >
        BRAND NAME
      </span>
    </div>
    <div
      class="sc-gKXOVf YVohq"
    >
      <div
        class="sc-gsnTZi avcHX"
      >
        <h1>
          NEW COLLECTION SHOES
        </h1>
        <span
          class="sc-iBkjds fkgdlf"
        >
          (
          3.3 of 5 start
          )
        </span>
        <div
          class="sc-hKMtZM hBbZGS"
        >
          <h3>
            DESCRIPTION
          </h3>
          <span
            class="sc-eCYdqJ jeXZll"
            data-testid="plus-minus-description"
          >
            +
          </span>
        </div>
        <span
          class="sc-jSMfEi dnCBFt"
        >
          <h3>
            MAKE A REVIEW
          </h3>
          <span
            class="sc-eCYdqJ jeXZll"
            data-testid="plus-minus-review"
          >
            -
          </span>
        </span>
        <div
          class="sc-kDDrLX iuImlR"
          data-testid="review-form"
        >
          <form
            class="sc-evZas gHCSVc"
            data-testid="form"
          >
            <label
              class="sc-fEOsli cItBsc"
            >
              Name:
              <input
                class="sc-bjUoiL faIaSM"
                name="name"
                placeholder="Name"
                required=""
                type="text"
                value=""
              />
            </label>
            <label
              class="sc-fEOsli cItBsc"
            >
              Email:
              <input
                class="sc-bjUoiL faIaSM"
                name="email"
                placeholder="Email"
                required=""
                type="email"
                value=""
              />
            </label>
            <label
              class="sc-fEOsli cItBsc"
            >
              Rating:
              <div>
                <svg
                  aria-hidden="true"
                  class="StyledIconBase-ea9ulj-0 lbJwfL sc-crXcEl eUFFki"
                  data-testid="ranking-0"
                  fill="currentColor"
                  focusable="false"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21.947 9.179a1.001 1.001 0 0 0-.868-.676l-5.701-.453-2.467-5.461a.998.998 0 0 0-1.822-.001L8.622 8.05l-5.701.453a1 1 0 0 0-.619 1.713l4.213 4.107-1.49 6.452a1 1 0 0 0 1.53 1.057L12 18.202l5.445 3.63a1.001 1.001 0 0 0 1.517-1.106l-1.829-6.4 4.536-4.082c.297-.268.406-.686.278-1.065z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="StyledIconBase-ea9ulj-0 lbJwfL sc-crXcEl eUFFki"
                  data-testid="ranking-1"
                  fill="currentColor"
                  focusable="false"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21.947 9.179a1.001 1.001 0 0 0-.868-.676l-5.701-.453-2.467-5.461a.998.998 0 0 0-1.822-.001L8.622 8.05l-5.701.453a1 1 0 0 0-.619 1.713l4.213 4.107-1.49 6.452a1 1 0 0 0 1.53 1.057L12 18.202l5.445 3.63a1.001 1.001 0 0 0 1.517-1.106l-1.829-6.4 4.536-4.082c.297-.268.406-.686.278-1.065z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="StyledIconBase-ea9ulj-0 lbJwfL sc-crXcEl eUFFki"
                  data-testid="ranking-2"
                  fill="currentColor"
                  focusable="false"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21.947 9.179a1.001 1.001 0 0 0-.868-.676l-5.701-.453-2.467-5.461a.998.998 0 0 0-1.822-.001L8.622 8.05l-5.701.453a1 1 0 0 0-.619 1.713l4.213 4.107-1.49 6.452a1 1 0 0 0 1.53 1.057L12 18.202l5.445 3.63a1.001 1.001 0 0 0 1.517-1.106l-1.829-6.4 4.536-4.082c.297-.268.406-.686.278-1.065z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="StyledIconBase-ea9ulj-0 lbJwfL sc-crXcEl eUFFki"
                  data-testid="ranking-3"
                  fill="currentColor"
                  focusable="false"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21.947 9.179a1.001 1.001 0 0 0-.868-.676l-5.701-.453-2.467-5.461a.998.998 0 0 0-1.822-.001L8.622 8.05l-5.701.453a1 1 0 0 0-.619 1.713l4.213 4.107-1.49 6.452a1 1 0 0 0 1.53 1.057L12 18.202l5.445 3.63a1.001 1.001 0 0 0 1.517-1.106l-1.829-6.4 4.536-4.082c.297-.268.406-.686.278-1.065z"
                  />
                </svg>
                <svg
                  aria-hidden="true"
                  class="StyledIconBase-ea9ulj-0 lbJwfL sc-crXcEl eUFFki"
                  data-testid="ranking-4"
                  fill="currentColor"
                  focusable="false"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M21.947 9.179a1.001 1.001 0 0 0-.868-.676l-5.701-.453-2.467-5.461a.998.998 0 0 0-1.822-.001L8.622 8.05l-5.701.453a1 1 0 0 0-.619 1.713l4.213 4.107-1.49 6.452a1 1 0 0 0 1.53 1.057L12 18.202l5.445 3.63a1.001 1.001 0 0 0 1.517-1.106l-1.829-6.4 4.536-4.082c.297-.268.406-.686.278-1.065z"
                  />
                </svg>
              </div>
            </label>
            <label
              class="sc-fEOsli cItBsc"
            >
              Title:
              <input
                class="sc-bjUoiL faIaSM"
                name="title"
                placeholder="Comment Title"
                required=""
                type="text"
                value=""
              />
            </label>
            <label
              class="sc-fEOsli cItBsc"
            >
              Comment:
              <textarea
                class="sc-idiyUo isBeHw"
                name="comment"
                placeholder="Comment"
                required=""
                rows="3"
              />
            </label>
            <p
              class="sc-breuTD fEGZDW"
              data-testid="error-msg"
            />
            <div
              class="sc-ksZaOG jRmgAk"
            >
              <input
                class="sc-hAZoDl dkwFTL"
                type="submit"
                value="Save review"
              />
              <button
                class="sc-fnykZs jvLVjt"
              >
                Clear review
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </section>
</div>
`;
